### PR TITLE
crypto: nrf_cc3xx v0.9.6

### PR DESCRIPTION
**Pull request by frkv**
-nrf_security: Fixing faulty configuration for number of entropy sources (for CTR_DRBG)
-nrf_security: Adding copy of mbedtls/cc3xx_kmu.h when cc3xx backend is enabled
-Adding nrf_cc3xx_platform/nrf_cc3xx_mbedcrypto v0.9.6
    -Adding version built against mbed TLS v2.24.0
    -Adding fix for regression in keys derived from KMU for ECB, CCM and GCM
    -Building ctr_drbg.c and entropy.c into nrf_cc3xx_platform library
    
manifest pr: https://github.com/nrfconnect/sdk-nrf/pull/3567

ref: NCSDK-7475
ref: NCSDK-7351
ref: NCSDK-7360
ref: NCSDK-7513

Signed-off-by: Torstein Grindvik <torstein.grindvik@nordicsemi.no>
Signed-off-by: Frank Audun Kvamtrø <frank.kvamtro@nordicsemi.no>

### CODE REVIEW

**Review by Vge0rge — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

File: nrf_security/configs/nrf-config.h.template [hunk 1]
```diff
@@ -1111,7 +1111,7 @@
  *      MBEDTLS_TLS_ECDH_RSA_WITH_CAMELLIA_128_GCM_SHA256
  *      MBEDTLS_TLS_ECDH_RSA_WITH_CAMELLIA_256_GCM_SHA384
  */
-#cmakedefine MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED
+//#cmakedefine MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED
```
Comment (tejlmand) [inline nrf_security/configs/nrf-config.h.template:1114]: This is wrong way to fix NCSDK-7513.

There are other users of nrf security, than openthread.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: nrf_security/configs/nrf-config.h.template [hunk 2]
```diff
@@ -3750,7 +3750,12 @@ it is (2^48 - 1), our restriction is :  (int - 0xFFFF - 0xF).*/
 typedef void mbedtls_ecp_restart_ctx;
 #endif
 
-
+/*
+ * Nordic added. Ensure that the MBEDTLS_SSL_SOME_SUITES_USE_TLS_CBC is enabled when PROTO_TLS is enabled.
+ */
+#if defined(MBEDTLS_SSL_PROTO_TLS1) || defined(MBEDTLS_SSL_PROTO_TLS1_1) || defined(MBEDTLS_SSL_PROTO_TLS1_2)
+        #define MBEDTLS_SSL_SOME_SUITES_USE_TLS_CBC
```
Comment (tejlmand) [inline nrf_security/configs/nrf-config.h.template:3757]: why is this needed ?
This template ALWAYS defines `MBEDTLS_SSL_PROTO_TLS1_2`: 
https://github.com/nrfconnect/sdk-nrfxlib/blob/8243ed209863ee1a3bf2a5d2735a939e8f890f4f/nrf_security/configs/nrf-config.h.template#L1647

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: nrf_security/configs/nrf-config.h.template (hunk 2 above)
Comment (tejlmand) [reply nrf_security/configs/nrf-config.h.template:3757]: sorry, wrong lien, i'm referring to the line above:
```
#if defined(MBEDTLS_SSL_PROTO_TLS1) || defined(MBEDTLS_SSL_PROTO_TLS1_1) || defined(MBEDTLS_SSL_PROTO_TLS1_2)
```

Comment (Vge0rge) [reply nrf_security/configs/nrf-config.h.template:3757]: Isn't that more future proof if later we want to use TLS1_3?

Comment (frkv) [reply nrf_security/configs/nrf-config.h.template:3757]: The changes added makes sense, please remove the "request for change" @tejlmand

File: nrf_security/configs/nrf-config.h.template [hunk 3]
```diff
@@ -3394,7 +3394,7 @@ it is (2^48 - 1), our restriction is :  (int - 0xFFFF - 0xF).*/
 #cmakedefine MBEDTLS_ECP_FIXED_POINT_OPTIM @MBEDTLS_ECP_FIXED_POINT_OPTIM@ /**< Enable fixed-point speed-up */
 
 /* Entropy options */
-//#define MBEDTLS_ENTROPY_MAX_SOURCES                1 /**< Maximum number of sources supported */
+#define MBEDTLS_ENTROPY_MAX_SOURCES                1 /**< Maximum number of sources supported */
```
Comment (tejlmand) [inline nrf_security/configs/nrf-config.h.template:3397]: alignment is off:
```suggestion
#define MBEDTLS_ENTROPY_MAX_SOURCES                  1 /**< Maximum number of sources supported */
```

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: nrf_security/configs/nrf-config.h.template (hunk 2 above)
Comment (tejlmand) [reply nrf_security/configs/nrf-config.h.template:3757]: in which way ?
If this is for future proofing if `TLS1_3` support, then the current:
https://github.com/nrfconnect/sdk-nrfxlib/blob/8243ed209863ee1a3bf2a5d2735a939e8f890f4f/nrf_security/configs/nrf-config.h.template#L1647

will probably become a:
```
 #cmakedefine MBEDTLS_SSL_PROTO_TLS1_2 
```
and in that case, the value for `#define MBEDTLS_SSL_SOME_SUITES_USE_TLS_CBC` could be a `cmakedefine` as well, so that it follows the pattern for the rest of the defines in this file.

Comment (frkv) [reply nrf_security/configs/nrf-config.h.template:3757]: Ah yes. That will be cleaner, agreed...

File: nrf_security/configs/nrf-config.h.template [hunk 4]
```diff
@@ -1644,7 +1644,18 @@
...
+#cmakedefine MBEDTLS_SSL_PROTO_TLS1_2
+
+/**
+ * \def MBEDTLS_SSL_SOME_SUITES_USE_TLS_CBC
+ *
+ * Enable the CBC construct used in TLS 1.2
+ *
+ * Requires: MBEDTLS_SSL_PROTO_TLS1_2
+ *
+ * Comment this macro to disable the CBC contrsuct for TLS 1.2
```
Comment (frkv) [inline nrf_security/configs/nrf-config.h.template:1656]: Typo: contrsuct

File: nrf_security/configs/nrf-config.h.template [hunk 5]
```diff
@@ -1644,7 +1644,18 @@
...
+ */
+#cmakedefine MBEDTLS_SSL_SOME_SUITES_USE_TLS_CBC
```
Comment (tejlmand) [inline nrf_security/configs/nrf-config.h.template:1658]: This seem to be Nordic addition, so should probably be moved to the bottom of this file and keep the previous comment:
```
* Nordic added.
```

**Comment by frkv** [comment]
There will be an additional commit coming promptly, but please start reviewing the latest changes @tejlmand

File: nrf_security/cmake/kconfig_mbedtls_configure.cmake [hunk 6]
```diff
@@ -50,6 +50,19 @@ macro(mbedtls_config_define_depends mbedtls_config)
   endforeach()
 endmacro()
 
+#
+# Internal macro which will enable the mbedtls configuration specified by the
+# first argument if at least one dependency is enabled
+#
+macro(mbedtls_config_define_depends_any mbedtls_config)
```
Comment (tejlmand) [inline nrf_security/cmake/kconfig_mbedtls_configure.cmake:57]: This is not being used anywhere.

Please wait on introducing such macros until there is a need.

File: nrf_security/configs/nrf-config.h.template [hunk 7]
```diff
@@ -2168,7 +2168,7 @@
  *            it, and considering stronger ciphers instead.
  *
  */
-#define MBEDTLS_ARC4_C
+#cmakedefine MBEDTLS_ARC4_C
```
Comment (tejlmand) [inline nrf_security/configs/nrf-config.h.template:2171]: `#cmakedefine` should only be used for settings we control using Kconfig + CMake.

I don't see ARC4 being configurable through Kconfig / CMake.

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: nrf_security/cmake/kconfig_mbedtls_configure.cmake (hunk 6 above)
Comment (Vge0rge) [reply nrf_security/cmake/kconfig_mbedtls_configure.cmake:57]: Ok then, I will remote that when the CI is finished

File: nrf_security/configs/nrf-config.h.template [hunk 8]
```diff
@@ -1178,7 +1178,7 @@
 //#define MBEDTLS_FS_IO
 
 /**
- * \def MBEDTLS_NO_DEFAULT_ENTROPY_SOURCES
+ * \def MBEDTLS_NO_DEFAULENTROPY_SOURCES
```
Comment (magnev) [inline nrf_security/configs/nrf-config.h.template:1181]: typo?

File: nrf_security/configs/nrf-config.h.template (hunk 2 above)
Comment (frkv) [reply nrf_security/configs/nrf-config.h.template]: This has been fixed

Comment (frkv) [reply nrf_security/configs/nrf-config.h.template]: This has been fixed

File: nrf_security/configs/nrf-config.h.template (hunk 3 above)
Comment (frkv) [reply nrf_security/configs/nrf-config.h.template]: This has been fixed

File: nrf_security/configs/nrf-config.h.template (hunk 7 above)
Comment (frkv) [reply nrf_security/configs/nrf-config.h.template:2171]: This has been fixed with adding section in cmake

File: nrf_security/configs/nrf-config.h.template (hunk 8 above)
Comment (frkv) [reply nrf_security/configs/nrf-config.h.template]: fixed

**Review by magnev — APPROVED** [review]

**Review by tejlmand — APPROVED** [review]

**Review by b-gent — APPROVED** [review]
